ps2-decompile/attach.c

```diff
@@ -3,188 +3,170 @@
+    }
+
+    if ((edesc->bEndpointAddress & 0x80) == 0) {
+        return -1;
+    }
+
+    if ((edesc->bmAttributes & 3) != 3) {
+        return -1;
+    }
+
+    if ((maxPacketSize = edesc->wMaxPacketSizeLB | (edesc->wMaxPacketSizeHB << 8)) >= 9) {
+        printf("wrong payload size %d\n", maxPacketSize);
+        return -1;
+    }
+
+    if ((g = unit_alloc(devId, maxPacketSize,idesc->bInterfaceNumber, idesc->bAlternateSetting)) == NULL) {
+        printf("unit_alloc failed\n");
+        return -1;
+    }
+
+    if ((g->configEnpoint = sceUsbdOpenPipe(devId, 0)) < 0) {
+        printf("sceUsbdOpenPipe failed(1)\n");
+        unit_free(g);
+        return -1;
+    }
+
+    printf("#### sceUsbdOpenPipe %d\n", g->configEndpoint);
+
+    if ((g->dataEndpoint = sceUsbdOpenPipe(devId, edesc)) < 0) {
+        printf("sceUsbdOpenPipe failed(2)\n");
+        unit_free(g);
+        return -1;
+    }
+
+    sceUsbdSetPrivateData(devId, g);
+    g->descriptorChecksPassed = 1;
+
+    if (cdesc->bLength != 9 ||
+        LOBYTE(cdesc->wTotalLength) != 0x22 || HIBYTE(cdesc->wTotalLength) ||
+        cdesc->bNumInterfaces != 1 || cdesc->bConfigurationValue != 1 ||
+        cdesc->iConfiguration || cdesc->bmAttributes != 0x80 ||
+        cdesc->maxPower != 0xC8) {
+        g->descriptorChecksPassed = 0;
+        printf("cdesc check failed\n");
+    }
+    if (hdesc->bLength != 9 ||
+        hdesc->bDescriptorType != 0x21 ||
+        LOBYTE(hdesc->bcdUSB) != 0x10 ||
+        HIBYTE(hdesc->bcdUSB) != 1 ||
+        hdesc->bDeviceClass ||
+        hdesc->bDeviceSubClass != 1 ||
+        hdesc->bDeviceProtocol != 0x22 ||
+        hdesc->bMaxPacketSize0 != 0x52 ||
+        LOBYTE(hdesc->idVendor)) {
+        g->descriptorChecksPassed = 0;
+        printf("hdesc check failed\n");
+    }
+    if (edesc->bEndpointAddress != 0x81 ||
+        edesc->bmAttributes != 3 ||
+        edesc->wMaxPacketSizeLB != 8 ||
+        edesc->wMaxPacketSizeHB ||
+        LOBYTE(edesc->bInterval) != 4) {
+        g->descriptorChecksPassed = 0;
+        printf("edesc check failed\n");
+    }
+
+    if ((getDeviceLocationRet = sceUsbdGetDeviceLocation(
+             devId, deviceLocationPath))) {
+        printf("usbmouse%d: sceUsbdGetDeviceLocation -> 0x%x\n", g->port1,
+               getDeviceLocationRet);
+        unit_free(g);
+        return -1;
+    }
+
+    printf("usbmouse%d: attached (port=", g->port1);
+
+    unrolled(0xFFFF8048, deviceLocationPath[0], v5, gp0);
+    unrolled(0xFFFF8044, deviceLocationPath[1], v5, gp0);
+    unrolled(0xFFFF8044, deviceLocationPath[2], v5, gp0);
+    unrolled(0xFFFF8044, deviceLocationPath[3], v5, gp0);
+    unrolled(0xFFFF8044, deviceLocationPath[4], v5, gp0);
+    unrolled(0xFFFF8044, deviceLocationPath[5], v5, gp0);
+    unrolled(0xFFFF8044, deviceLocationPath[6], v5, gp0);
+
+    printf((const char *)(gp0 + 0xFFFF804C));
+    printf((const char *)(gp2 + 0xFFFF8020));
+    printf((const char *)(gp1 + 0xFFFF8050),deviceLocationPath[0]);
```

Comment:
This should just be a loop. `deviceLocationPath[0]` is more useful as an integer than a character, so `deviceLocationPath` should not be printed with `%s`.

Reply:
in that case it would be more sane to unroll it since 0 being a special case would be more like so.
```C
    printf("Device Port: %i", deviceLocationPath[0]);
```
This has been done, working on other issues atm.